### ISSUE TITLE
Improve Private Live Chat (IM module) 

### ISSUE THREAD

**Issue by pH-7** [pH7Software/pH7-Social-Dating-CMS#138]
The Instant Messenger ("IM" module) needs improvements, better responsive design (the chat window doesn't fit well on mobiles) and fix some bugs (sometimes users don't get the message directly or the alert sound doesn't work).

**Comment by polynamaude** [comment]
I've made a chat module for Joomla and I could integrate it pretty easily with pH7CMS.
Only one question, do I make a new module or modify the actual one if you want to keep it possible to use a external service ?

Have a nice Week End!

**Comment by pH-7** [comment]
Oh, that's awesome! Maybe the best would be to create a new module, ie "native-chat", "advanced-chat", etc. (later on, we can still remove the old chat module ;)).

Have a great weekend, too! (sorry, just see your message now).

**Comment by polynamaude** [comment]
Hi !
Thanks for the answer, there's no such problem as the time it took as we're all pretty much giving as much we can. It's impossible to do more than what's possible !

I'll take a look at how does the AJAX API works, I understood that there's already a base done. In my actual solution I simply use jQuery and some options sent to Joomla (task, view, tmpl=component). I'll have to find out how to do all this with pH7 but that's a great project. I'm currently job-seeking and that will be making part of my portfolio. Last year I got in a bad mood and destroyed many backups including lots of code I wrote. As I can't really ask passed employer to send me back what they paid me to do then I make new project.

I'll get you back on that, the code is pretty easy to understand and follow most common design pattern so I shouldn't have too much problem finding answer by myself.

Ever heard about the Therrac-25 (or a similar number) software bug that killed (!!!) half a dozen of persons. It was the software controlling a irradiation therapy device sold by Atomic Energy Canada Limited. They re-used the software on a newer device that didn't have some hardware based protection preventing from using high power without a beam splitter. The error only happened when the operator became fast enough to cause a race condition between the test shot and the real shot. It's after this that the federal gov't imposed software code review and testing for medical devices. Ain't that great as it now makes us able to read code that other folks have written. Maybe that without this event we wouldn't even have code review and design pattern !

Have a nice Week-End on the old continent ;-)

**Comment by polynamaude** [comment]
Hi (again) !

Can I presume that the CMS validate the current user session (to prevent XSS/hijacking). ?

Where do I add database table creation ? Is there (probably yes) a standard place where it's all done ? I can do it in the module bootstrap ?

That should be it with questions (for now)!

Thanks !

Le 23 avril 2018 15:04:31 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Hey Polyna-Maude!
>
>Sure thing!
>On pH7CMS structure, you should create `assets/ajax/` folder in your
>module (e.g.
>[https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/im/assets/ajax/MessengerAjax.php](assets/ajax/MessengerAjax.php)
>for the IM module) and create a public folder (where you will have the
>JS file) in the theme folder of the module (e.g.
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/templates/system/modules/im/themes/base/js/Messenger.js)
>and you do the request like the following
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/templates/system/modules/im/themes/base/js/Messenger.js#L257
>
>Then, ajax request is managed by the FrontController
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Mvc/Router/FrontController.class.php#L438-L459
>
>Hopefully, this will help you a little bit more ;)
>
>I'm very glad if you can make your portfolio bigger at the same time :)
>
>Let me know if you have any other questions! :)
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-383687073

-- 
Polyna-Maude R.-Summerside

**Comment by pH-7** [comment]
Hey PM,

Haha, yeah ^^ I even wrote a post a while ago explaining the name https://ph7cms.com/why-ph7cms-name/ :D

No worries! I totally understand! Be yourself when you answer to people :)

Concerning the user session, below is the check that makes sure if a user is logged in or not.
https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/core/classes/UserCore.php#L57-L60
So, basically it checks also if the client IP address and the HTTP user agent haven't changed during the session.

**Comment by polynamaude** [comment]
Hi Ph !

Thanks, it's exactly what I understood by reading the source code.

I'm quite surprised that you wrote all that on your own. I just found the tests folder ! Great, I'll take a dive to check it out.

I prefer to understand how it goes than to do some try - errors loop that are a waste of time. But for now I get less problem to understand how the piece fit together than what I had for many PHP projects.

Thanks for the support !
PM

Le 24 avril 2018 14:04:03 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Hey PM,
>
>Haha, yeah ^^ I even wrote a post a while ago explaining the name
>https://ph7cms.com/why-ph7cms-name/ :D
>
>No worries! I totally understand! Be yourself when you answer to people
>:)
>
>Concerning the user session, below is the check that makes sure if a
>user is logged in or not.
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/core/classes/UserCore.php#L57-L60
>So, basically it checks also if the client IP address and the HTTP user
>agent haven't changed during the session.
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-384026332

-- 
Polyna-Maude R.-Summerside

**Comment by pH-7** [comment]
Thanks PM! Ah.., working a little every day, gives a lot after a few years (it's like a book :D). And when you are driven by passion, time doesn't count 😀 

Yes, I've been added tests for a little while only 😢 So, this is also something that has to be improved! I'm far from the 100% test coverage.


You're very welcome!

**Comment by polynamaude** [comment]
Little question...
Got my answer for database table on modules installation and configuration.

But here's the question...
Is there a easy line I can use to understand what belong to framework and what is in the cms side? Can I make the assumption that everything not "dating related" is on the framework?

More I get into this and the more I find interesting what's done. Pretty lightweight and fast. Can be a great replacement for Cake or such, just get some outside library for what's needed like nudity detector in pH7.

Le 25 avril 2018 15:52:45 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Thanks PM! Ah.., working a little every day, gives a lot after a few
>years (it's like a book :D). And when you are driven by passion, time
>doesn't count 😀 
>
>Yes, I've been added tests for a little while only 😢 So, this is also
>something that has to be improved! I'm far from the 100% test coverage.
>
>
>You're very welcome!
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-384413122

-- 
Polyna-Maude R.-Summerside

**Comment by pH-7** [comment]
Hey PM!

Sorry, I just released I skipped your question concerning the database.

The plan is that normally you add "install/sql/MySQL/install.sql" inside of your module.
Although the SQL files are empty, you can see an example in the honeypot module https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_repository/module/fake-admin-panel/install/sql/MySQL/

And then, those SQL files are run in https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/admin123/inc/class/Module.php#L202-L213
(but, that sql part hasn't been used/tested.. yet because of lack of time).

FYI, when the module is located in https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_repository/module
it will appear in Admin Panel -> Mod -> 3rd-party Mods Manager
Then, once you install it, it will be moved to https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_protected/app/modules

Most (if not all) of the packages in the framework aren't "dating" or "social" related. Now, unfortunately, several packages are related with others and aren't "that" really standalone for the moment and some parts sadly depend with the CMS part...
That's something that definitely needs to be improved. And with more dependency injections, etc.

**Comment by pH-7** [comment]
Hey PM! Sorry for my late reply :)

Great for the changes! Don't worry. The best, it's just to create a new Pull Request and I will be pleased to review your changes!

You're right, I was using Geany in the past for years, then Subline, then Atom and now I'm only using PHPStorm. It's by far the best. Once you start using it, you can't use another one. It will save you alot of time.

Yeap, could be a good option. You can easily display a "Sorry" page like how it has been done for maintenance page and the other ones in https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php and create a HTML page in https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_protected/app/system/global/views/base/tpl/other
or just reuse the [https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L70](banned page) (maybe just slightly change the HTML message there...).
Yes, you can send a 403 header (like what the banned page already does https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L73

:+1:

Indeed! I mean, it wasn't that big on day one... For pH7CMS, I didn't use that often pen and paper neither. At work, I use [BDD](https://en.wikipedia.org/wiki/Behavior-driven_development) for some features (the PHP Behat framework is nice for that). And create user stories specs which can help you depending on the complexity.

Have a great day too! Hopefully this will help!

**Comment by polynamaude** [comment]
Hi PH,

I've just downloaded PHPStorm, I'll give it a try later on today.

I've had to take a look at what does BDD meant and it's something I did already heard of for Business Process Development (using Eclipse Business Process Modeling). Sounds great to me as a way to work closer to the real knowledge of the business (or other type) process. As it says "object oriented programming language doesn't mean object oriented development". Too often we think in how we're gona code it before looking at the real design of the complete information flow.

Do you know a way I can download (sync) the GitHub issue ?

Do you still want to use SASS/LESS for the CMS? I'm working on a system for use with Joomla, on the fly compilation (and caching). If so I could send you some of my work and we could try to integrate into the existing caching/compression system. I want to limit the dependency so no Ruby (even if nice), no node server or java. Only PHP based solutions.

Did I read somewhere that there's a Twig support ? Or maybe I'm just messed up and your templating use a Twig-Like markup !?

PS: We should work a way to make the Captcha configurable on a level of difficulty. Look at mine is nearly impossible to use! (www.polynamaude.com)

Have a nice Week-End !

Le 3 mai 2018 14:14:05 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Hey PM! Sorry for my late reply :)
>
>Great for the changes! Don't worry. The best, it's just to create a new
>Pull Request and I will be pleased to review your changes!
>
>You're right, I was using Geany in the past for years, then Subline,
>then Atom and now I'm only using PHPStorm. It's by far the best. Once
>you start using it, you can't use another one. It will save you alot of
>time.
>
>Yeap, could be a good option. You can easily display a "Sorry" page
>like how it has been done for maintenance page and the other ones in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php
>and create a HTML page in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_protected/app/system/global/views/base/tpl/other
>or just reuse the
>[https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L70](banned
>page) (maybe just slightly change the HTML message there...).
>Yes, you can send a 403 header (like what the banned page already does
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L73
>
>:+1:
>
>Indeed! I mean, it wasn't that big on day one... For pH7CMS, I didn't
>use that often pen and paper neither. At work, I use
>[BDD](https://en.wikipedia.org/wiki/Behavior-driven_development) for
>some features (the PHP Behat framework is nice for that). And create
>user stories specs which can help you depending on the complexity.
>
>Have a great day too! Hopefully this will help!
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-386387979

-- 
Polyna-Maude R.-Summerside

**Comment by polynamaude** [comment]
Hi !
Maybe you got a "already" did before idea but how do I mark user getting offline ? Was thinking of a "keepalive" by Ajax every 30 second. If the software can't find one for a designated user then consider the user is out of chat module. Do a cleanup by cron at midnight or so.

The other option would be to keep message with undelivered status and purge them after a delay that the user hasn't connected. I don't want excess load so I try to keep table light. By your experience, what do you suggest ?

PM

Le 3 mai 2018 14:14:05 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Hey PM! Sorry for my late reply :)
>
>Great for the changes! Don't worry. The best, it's just to create a new
>Pull Request and I will be pleased to review your changes!
>
>You're right, I was using Geany in the past for years, then Subline,
>then Atom and now I'm only using PHPStorm. It's by far the best. Once
>you start using it, you can't use another one. It will save you alot of
>time.
>
>Yeap, could be a good option. You can easily display a "Sorry" page
>like how it has been done for maintenance page and the other ones in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php
>and create a HTML page in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_protected/app/system/global/views/base/tpl/other
>or just reuse the
>[https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L70](banned
>page) (maybe just slightly change the HTML message there...).
>Yes, you can send a 403 header (like what the banned page already does
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L73
>
>:+1:
>
>Indeed! I mean, it wasn't that big on day one... For pH7CMS, I didn't
>use that often pen and paper neither. At work, I use
>[BDD](https://en.wikipedia.org/wiki/Behavior-driven_development) for
>some features (the PHP Behat framework is nice for that). And create
>user stories specs which can help you depending on the complexity.
>
>Have a great day too! Hopefully this will help!
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-386387979

-- 
Polyna-Maude R.-Summerside

**Comment by polynamaude** [comment]
Hi PH!

Sent you some code by email but need some fixing. Forgot some slashes in RegExp and a call in the class constructor. As I explained, it's a fast solution but I'll make better when I find a nice way to hook into the controller. Maybe we should move the validation code for banned IP earlier in the process (before calling FrontController/RunRouter but you know better than me the dependency.

Have a great week end !

PM

Le 3 mai 2018 14:14:05 EDT, "♚ PH⑦ Soria ♛" <notifications@github.com> a écrit :
>Hey PM! Sorry for my late reply :)
>
>Great for the changes! Don't worry. The best, it's just to create a new
>Pull Request and I will be pleased to review your changes!
>
>You're right, I was using Geany in the past for years, then Subline,
>then Atom and now I'm only using PHPStorm. It's by far the best. Once
>you start using it, you can't use another one. It will save you alot of
>time.
>
>Yeap, could be a good option. You can easily display a "Sorry" page
>like how it has been done for maintenance page and the other ones in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php
>and create a HTML page in
>https://github.com/pH7Software/pH7-Social-Dating-CMS/tree/master/_protected/app/system/global/views/base/tpl/other
>or just reuse the
>[https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L70](banned
>page) (maybe just slightly change the HTML message there...).
>Yes, you can send a 403 header (like what the banned page already does
>https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Page/Page.class.php#L73
>
>:+1:
>
>Indeed! I mean, it wasn't that big on day one... For pH7CMS, I didn't
>use that often pen and paper neither. At work, I use
>[BDD](https://en.wikipedia.org/wiki/Behavior-driven_development) for
>some features (the PHP Behat framework is nice for that). And create
>user stories specs which can help you depending on the complexity.
>
>Have a great day too! Hopefully this will help!
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pH7Software/pH7-Social-Dating-CMS/issues/138#issuecomment-386387979

-- 
Polyna-Maude R.-Summerside

**Comment by pH-7** [comment]
Thanks for the suggestion.

I reviewed this during backlog cleanup. This request is valid but outside the current 18.x delivery scope, so I’m closing it as not planned for now to keep the tracker focused on active work.

If you want to push this forward, please open a new proposal with a minimal implementation plan (scope, DB/API impact, migration notes, and backward-compatibility expectations).

Take care and thank you for using pH7 Social Dating Builder,

[Pierre-Henry](https://github.com/pH-7)
